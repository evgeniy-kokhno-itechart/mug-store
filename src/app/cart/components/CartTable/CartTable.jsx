import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { PropTypes } from 'prop-types';
import { ResponsiveEllipsis, Table, ItemCounter } from '../../../shared';
import ProductPrice from '../ProductPrice';
import './CartTable.scss';

class CartTable extends Component {
  columns = [
    {
      key: 'image',
      label: '',
      content: product => (
        <Link to={`/products/${product.id}`} className="clickable ">
          <img src={product.imageURL} alt={product.title} className="product-image" />
        </Link>
      ),
      customClasses: 'cart-table--column-image',
    },

    {
      path: 'title',
      label: 'Title',
      content: product => (
        <Link to={`/products/${product.id}`} className="clickable product-title">
          {product.title}
        </Link>
      ),
      customClasses: 'cart-table--column-title',
    },

    {
      path: 'description',
      label: 'Details',
      content: product => (
        <ResponsiveEllipsis
          text={product.description}
          maxLine="2"
          ellipsis="..."
          basedOn="words"
          customClasses="product-description"
        />
      ),
      customClasses: 'cart-table--column-details',
    },

    {
      key: 'quantity',
      label: 'Quantity',
      content: product => (
        <ItemCounter
          item={product}
          onQuantityChanged={this.props.changeQuantity}
        />
      ),
      customClasses: 'cart-table--column-quantity',
    },

    {
      path: 'cost',
      label: 'Cost',
      content: product => <ProductPrice price={product.currentCurrencyCost} isCurrencyLoading={this.props.isCurrencyLoading} />,
      customClasses: 'text-center cart-table--column-cost',
    },

    {
      key: 'deleteFromCart',
      content: product => (
        <button type="button" id={product.id} className="button button-solid" onClick={this.handleDeleteProductFromCart}>
          <FontAwesomeIcon icon="trash" />
        </button>
      ),
      customClasses: 'cart-table--column-delete',
    },
  ];

  handleDeleteProductFromCart = (e) => {
    this.props.deleteProductFromCart(e.target.id);
  };

  render() {
    const { productsInCart, sortColumn } = this.props;
    return <Table columns={this.columns} sortColumn={sortColumn} items={productsInCart} customClasses="cart-table" />;
  }
}

CartTable.propTypes = {
  isCurrencyLoading: PropTypes.bool.isRequired,

  productsInCart: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      imageURL: PropTypes.string,
      title: PropTypes.string,
      description: PropTypes.string,
      currentCurrencyCost: PropTypes.number,
      quantity: PropTypes.number,
    }),
  ).isRequired,
  sortColumn: PropTypes.string.isRequired,
  changeQuantity: PropTypes.func.isRequired,
  deleteProductFromCart: PropTypes.func.isRequired,
};

export default CartTable;
