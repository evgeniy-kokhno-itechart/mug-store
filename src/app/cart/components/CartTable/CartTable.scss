@import "Mixins";

.cart-table {
  margin: 0 1.5rem 1.5rem 1.5rem;
  &__column-image {
    width: 12%;
  }

  &__column-title {
    width: 10%;
  }

  &__column-details {
    width: 58%;

    @include respond(tab-portrait) {
      width: 30%;
    }

    @include respond(phone) {
      display: none;
    }
  }

  &__column-quantity {
    width: 7%;
  }

  &__column-cost {
    width: 7%;
    text-align: center;
  }

  &__column-delete {
    width: 5%;
  }
}