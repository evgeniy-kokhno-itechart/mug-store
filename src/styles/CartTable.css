.cart-table--column__image {
  width: 15%;
}
.cart-table--column__title {
  width: 10%;
}
.cart-table--column__details {
  width: 58%;
}
.cart-table--column__quantity {
  width: 5%;
}
.cart-table--column__cost {
  width: 5%;
}
.cart-table--column__delete {
  width: 5%;
}
