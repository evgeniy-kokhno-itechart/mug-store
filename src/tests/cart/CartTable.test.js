/* eslint-disable no-undef */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import CartTable from '../../app/cart/components/CartTable';

configure({ adapter: new Adapter() });

describe('<CartTable />', () => {
  let CartTableWrapper;
  let changeQuantity;
  let deleteProductFromCart;

  beforeEach(() => {
    changeQuantity = jest.fn();
    deleteProductFromCart = jest.fn();

    CartTableWrapper = shallow(
      <CartTable
        isCurrencyLoading
        productsInCart={[
          {
            id: '1',
            imageURL: '#',
            title: 'test product 1',
            description: 'test product 1 description',
            currentCurrencyPrice: 5,
            quantity: 3,
          },
          {
            id: '2',
            imageURL: '#',
            title: 'test product 2',
            description: 'test product 2 description',
            currentCurrencyPrice: 10,
            quantity: 5,
          },
        ]}
        sortColumn="id"
        changeQuantity={changeQuantity}
        deleteProductFromCart={deleteProductFromCart}
      />,
    );
  });

  it('renders properly with default props', () => {
    expect(CartTableWrapper).toMatchSnapshot();
  });
});
