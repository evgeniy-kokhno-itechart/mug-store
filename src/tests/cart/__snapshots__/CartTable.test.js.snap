// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartTable /> renders properly with default props 1`] = `
<Table
  columns={
    Array [
      Object {
        "content": [Function],
        "customClasses": "cart-table--column__image",
        "key": "image",
        "label": "",
      },
      Object {
        "content": [Function],
        "customClasses": "cart-table--column__title",
        "label": "Title",
        "path": "title",
      },
      Object {
        "content": [Function],
        "customClasses": "cart-table--column__details",
        "label": "Details",
        "path": "description",
      },
      Object {
        "content": [Function],
        "customClasses": "cart-table--column__quantity",
        "key": "quantity",
        "label": "Quantity",
      },
      Object {
        "content": [Function],
        "customClasses": "text-center cart-table--column__cost",
        "label": "Cost",
        "path": "cost",
      },
      Object {
        "content": [Function],
        "customClasses": "cart-table--column__delete",
        "key": "deleteFromCart",
      },
    ]
  }
  headerExcluded={false}
  items={
    Array [
      Object {
        "currentCurrencyPrice": 5,
        "description": "test product 1 description",
        "id": "1",
        "imageURL": "#",
        "quantity": 3,
        "title": "test product 1",
      },
      Object {
        "currentCurrencyPrice": 10,
        "description": "test product 2 description",
        "id": "2",
        "imageURL": "#",
        "quantity": 5,
        "title": "test product 2",
      },
    ]
  }
  sortColumn="id"
/>
`;
